// @tarojs/components-rn/dist/components/Radio/index.js
/**
 * 半受控组件
 *
 * ✔ value
 * ✔ disabled
 * ✔ checked
 * ✔ color
 * ✔ onChange
 */
import * as React from 'react';
import { View, TouchableWithoutFeedback, Platform } from 'react-native';
import Icon from '../Icon';
import styles from './styles';
import { noop } from '../../utils';
class _Radio extends React.Component {
    constructor() {
        super(...arguments);
        this.$touchable = React.createRef();
        this.state = {
            checked: false
        };
        this._simulateNativePress = (evt) => {
            const node = this.$touchable.current;
            node && node.touchableHandlePress(evt);
        };
        this.onPress = () => {
            const { disabled, onChange = noop, value } = this.props;
            if (disabled)
                return;
            if (this.state.checked)
                return;
            onChange({
                value,
                checked: !this.state.checked
            });
            this.setState({ checked: !this.state.checked });
        };
    }
    static getDerivedStateFromProps(props, state) {
        return props.checked !== state.checked ? {
            checked: !!props.checked
        } : null;
    }
    render() {
        const { style, color, } = this.props;
        const isChecked = this.state.checked;
        const iconSize = Platform.OS === 'ios' ? 24 : 21;
        return (React.createElement(TouchableWithoutFeedback, { onPress: this.onPress, ref: this.$touchable },
            React.createElement(View, { style: [
                    styles.wrapper,
                    isChecked && styles.wrapperChecked,
                    isChecked && { borderColor: color },
                    style
                ] },
                React.createElement(Icon, { type: "success", size: iconSize, color: color, style: [styles.wrapperIcon, isChecked && styles.wrapperCheckedIcon] }))));
    }
}
_Radio.defaultProps = {
    value: '',
    color: '#09BB07',
};
export default _Radio;
