// @tarojs/components-rn/dist/components/Checkbox/index.js
/**
 * 半受控组件
 *
 * ✔ value
 * ✔ disabled
 * ✔ checked
 * ✔ color
 * ✔ onChange
 *
 * @see https://wechat.design/brand/color
 */
import * as React from 'react';
import { View, TouchableWithoutFeedback } from 'react-native';
import Icon from '../Icon';
import styles from './styles';
import { noop } from '../../utils';
class _Checkbox extends React.Component {
    constructor() {
        super(...arguments);
        this.$touchable = React.createRef();
        this.state = {
            checked: false
        };
        this._simulateNativePress = (evt) => {
            const node = this.$touchable.current;
            node && node.touchableHandlePress(evt);
        };
        this.onPress = () => {
            const { disabled, onChange = noop, value } = this.props;
            if (disabled)
                return;
            onChange({
                value,
                checked: !this.state.checked
            });
            this.setState({ checked: !this.state.checked });
        };
    }
    static getDerivedStateFromProps(props, state) {
        return props.checked !== state.checked ? {
            checked: !!props.checked
        } : null;
    }
    render() {
        const { style, color, } = this.props;
        return (React.createElement(TouchableWithoutFeedback, { onPress: this.onPress, ref: this.$touchable },
            React.createElement(View, { style: [styles.wrapper, style, this.state.checked && styles.wrapperChecked] },
                React.createElement(Icon, { type: "success_no_circle", size: 18, color: color, style: [styles.wrapperIcon, this.state.checked && styles.wrapperCheckedIcon] }))));
    }
}
_Checkbox.defaultProps = {
    value: '',
    color: '#09BB07',
};
export default _Checkbox;
