// @tarojs/mini-runner/dist/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const fs = require("fs-extra");
const resolvePath = require("resolve");
const t = require("babel-types");
const lodash_1 = require("lodash");
const chalk_1 = require("chalk");
const constants_1 = require("./constants");
exports.isNodeModule = (filename) => constants_1.NODE_MODULES_REG.test(filename);
function isNpmPkg(name) {
    if (/^(\.|\/)/.test(name)) {
        return false;
    }
    return true;
}
exports.isNpmPkg = isNpmPkg;
function isQuickAppPkg(name) {
    return /^@(system|service)\.[a-zA-Z]{1,}/.test(name);
}
exports.isQuickAppPkg = isQuickAppPkg;
function isEmptyObject(obj) {
    if (obj == null) {
        return true;
    }
    for (const key in obj) {
        if (obj.hasOwnProperty(key)) {
            return false;
        }
    }
    return true;
}
exports.isEmptyObject = isEmptyObject;
function traverseObjectNode(node, buildAdapter, parentKey) {
    if (node.type === 'ClassProperty' || node.type === 'ObjectProperty') {
        const properties = node.value.properties;
        const obj = {};
        properties.forEach(p => {
            let key = t.isIdentifier(p.key) ? p.key.name : p.key.value;
            if (constants_1.CONFIG_MAP[buildAdapter][key] === false) {
                return;
            }
            if (parentKey !== 'usingComponents' && constants_1.CONFIG_MAP[buildAdapter][key]) {
                key = constants_1.CONFIG_MAP[buildAdapter][key];
            }
            obj[key] = traverseObjectNode(p.value, buildAdapter, key);
        });
        return obj;
    }
    if (node.type === 'ObjectExpression') {
        const properties = node.properties;
        const obj = {};
        properties.forEach(p => {
            let key = t.isIdentifier(p.key) ? p.key.name : p.key.value;
            if (constants_1.CONFIG_MAP[buildAdapter][key] === false) {
                return;
            }
            if (parentKey !== 'usingComponents' && constants_1.CONFIG_MAP[buildAdapter][key]) {
                key = constants_1.CONFIG_MAP[buildAdapter][key];
            }
            obj[key] = traverseObjectNode(p.value, buildAdapter, key);
        });
        return obj;
    }
    if (node.type === 'ArrayExpression') {
        return node.elements.map(item => traverseObjectNode(item, buildAdapter));
    }
    if (node.type === 'NullLiteral') {
        return null;
    }
    return node.value;
}
exports.traverseObjectNode = traverseObjectNode;
function isAliasPath(name, pathAlias = {}) {
    const prefixs = Object.keys(pathAlias);
    if (prefixs.length === 0) {
        return false;
    }
    return prefixs.includes(name) || (new RegExp(`^(${prefixs.join('|')})/`).test(name));
}
exports.isAliasPath = isAliasPath;
function replaceAliasPath(filePath, name, pathAlias = {}) {
    // 后续的 path.join 在遇到符号链接时将会解析为真实路径，如果
    // 这里的 filePath 没有做同样的处理，可能会导致 import 指向
    // 源代码文件，导致文件被意外修改
    filePath = fs.realpathSync(filePath);
    const prefixs = Object.keys(pathAlias);
    if (prefixs.includes(name)) {
        return promoteRelativePath(path.relative(filePath, fs.realpathSync(resolveScriptPath(pathAlias[name]))));
    }
    const reg = new RegExp(`^(${prefixs.join('|')})/(.*)`);
    name = name.replace(reg, function (m, $1, $2) {
        return promoteRelativePath(path.relative(filePath, path.join(pathAlias[$1], $2)));
    });
    return name;
}
exports.replaceAliasPath = replaceAliasPath;
function promoteRelativePath(fPath) {
    const fPathArr = fPath.split(path.sep);
    let dotCount = 0;
    fPathArr.forEach(item => {
        if (item.indexOf('..') >= 0) {
            dotCount++;
        }
    });
    if (dotCount === 1) {
        fPathArr.splice(0, 1, '.');
        return fPathArr.join('/');
    }
    if (dotCount > 1) {
        fPathArr.splice(0, 1);
        return fPathArr.join('/');
    }
    return fPath.replace(/\\/g, '/');
}
exports.promoteRelativePath = promoteRelativePath;
function resolveScriptPath(p) {
    const realPath = p;
    const taroEnv = process.env.TARO_ENV;
    const SCRIPT_EXT = constants_1.JS_EXT.concat(constants_1.TS_EXT);
    for (let i = 0; i < SCRIPT_EXT.length; i++) {
        const item = SCRIPT_EXT[i];
        if (taroEnv) {
            if (fs.existsSync(`${p}.${taroEnv}${item}`)) {
                return `${p}.${taroEnv}${item}`;
            }
            if (fs.existsSync(`${p}${path.sep}index.${taroEnv}${item}`)) {
                return `${p}${path.sep}index.${taroEnv}${item}`;
            }
            if (fs.existsSync(`${p.replace(/\/index$/, `.${taroEnv}/index`)}${item}`)) {
                return `${p.replace(/\/index$/, `.${taroEnv}/index`)}${item}`;
            }
        }
        if (fs.existsSync(`${p}${item}`)) {
            return `${p}${item}`;
        }
        if (fs.existsSync(`${p}${path.sep}index${item}`)) {
            return `${p}${path.sep}index${item}`;
        }
    }
    return realPath;
}
exports.resolveScriptPath = resolveScriptPath;
function buildUsingComponents(filePath, sourceDir, pathAlias, components, isComponent) {
    const usingComponents = Object.create(null);
    for (const component of components) {
        let componentPath = component.path;
        if (isAliasPath(componentPath, pathAlias)) {
            componentPath = replaceAliasPath(filePath, componentPath, pathAlias);
        }
        componentPath = resolveScriptPath(path.resolve(filePath, '..', componentPath));
        if (fs.existsSync(componentPath)) {
            if (constants_1.NODE_MODULES_REG.test(componentPath) && !constants_1.NODE_MODULES_REG.test(filePath)) {
                componentPath = componentPath.replace(constants_1.NODE_MODULES_REG, path.join(sourceDir, 'npm'));
            }
            componentPath = promoteRelativePath(path.relative(filePath, componentPath));
        }
        else {
            componentPath = component.path;
        }
        if (component.name) {
            const componentName = component.name.split('|')[0];
            usingComponents[componentName] = componentPath.replace(path.extname(componentPath), '');
        }
    }
    return Object.assign({}, isComponent ? { component: true } : { usingComponents: {} }, components.length ? {
        usingComponents
    } : {});
}
exports.buildUsingComponents = buildUsingComponents;
const npmCached = {};
function resolveNpmSync(pkgName, root) {
    try {
        if (!npmCached[pkgName]) {
            return resolvePath.sync(pkgName, { basedir: root });
        }
        return npmCached[pkgName];
    }
    catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
            throw new Error(`包 ${pkgName} 未安装`);
        }
        return null;
    }
}
exports.resolveNpmSync = resolveNpmSync;
function recursiveMerge(src, ...args) {
    return lodash_1.mergeWith(src, ...args, (value, srcValue) => {
        const typeValue = typeof value;
        const typeSrcValue = typeof srcValue;
        if (typeValue !== typeSrcValue)
            return;
        if (Array.isArray(value) && Array.isArray(srcValue)) {
            return value.concat(srcValue);
        }
        if (typeValue === 'object') {
            return recursiveMerge(value, srcValue);
        }
    });
}
exports.recursiveMerge = recursiveMerge;
function getInstalledNpmPkgPath(pkgName, basedir) {
    const resolvePath = require('resolve');
    try {
        return resolvePath.sync(`${pkgName}/package.json`, { basedir });
    }
    catch (err) {
        return null;
    }
}
exports.getInstalledNpmPkgPath = getInstalledNpmPkgPath;
function printLog(type, tag, filePath) {
    const typeShow = constants_1.processTypeMap[type];
    const tagLen = tag.replace(/[\u0391-\uFFE5]/g, 'aa').length;
    const tagFormatLen = 8;
    if (tagLen < tagFormatLen) {
        const rightPadding = new Array(tagFormatLen - tagLen + 1).join(' ');
        tag += rightPadding;
    }
    const padding = '';
    filePath = filePath || '';
    if (typeof typeShow.color === 'string') {
        console.log(chalk_1.default[typeShow.color](typeShow.name), padding, tag, padding, filePath);
    }
    else {
        console.log(typeShow.color(typeShow.name), padding, tag, padding, filePath);
    }
}
exports.printLog = printLog;
function removeHeadSlash(str) {
    return str.replace(/^(\/|\\)/, '');
}
exports.removeHeadSlash = removeHeadSlash;
function npmCodeHack(filePath, content, buildAdapter) {
    // 修正core-js目录 _global.js
    // 修正所有用到过lodash的第三方包
    // 注：@tarojs/taro-alipay/dist/index.js,@tarojs/taro/dist/index.esm.js里面也有lodash相关的代码
    content = content && content.replace(/(\|\||:)\s*Function\(['"]return this['"]\)\(\)/g, function (match, first, second) {
        return `${first} ${constants_1.GLOBAL_PROPS}`;
    });
    const basename = path.basename(filePath);
    switch (basename) {
        case 'mobx.js':
            // 解决支付宝小程序全局window或global不存在的问题
            content = content.replace(/typeof window\s{0,}!==\s{0,}['"]undefined['"]\s{0,}\?\s{0,}window\s{0,}:\s{0,}global/, 'typeof window !== "undefined" ? window : typeof global !== "undefined" ? global : {}');
            break;
        case '_html.js':
            content = 'module.exports = false;';
            break;
        case '_microtask.js':
            content = content.replace('if(Observer)', 'if(false && Observer)');
            // IOS 1.10.2 Promise BUG
            content = content.replace('Promise && Promise.resolve', 'false && Promise && Promise.resolve');
            break;
    }
    if (buildAdapter === "alipay" /* ALIPAY */ && content.replace(/\s\r\n/g, '').length <= 0) {
        content = '// Empty file';
    }
    return content;
}
exports.npmCodeHack = npmCodeHack;
