// @tarojs/mini-runner/dist/utils/logHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = require("chalk");
const ora = require("ora");
const fp_1 = require("lodash/fp");
const formatMessages = require("webpack-format-messages");
// const syntaxErrorLabel = 'Syntax error:'
const LOG_MAP = {
    ["weapp" /* WEAPP */]: {
        OPEN: 'è¯·æ‰“å¼€å¾®ä¿¡å°ç¨‹åºå¼€å‘è€…å·¥å…·è¿›è¡ŒæŸ¥çœ‹'
    },
    ["alipay" /* ALIPAY */]: {
        OPEN: 'è¯·æ‰“å¼€æ”¯ä»˜å®å°ç¨‹åºå¼€å‘è€…å·¥å…·è¿›è¡ŒæŸ¥çœ‹'
    },
    ["qq" /* QQ */]: {
        OPEN: 'è¯·æ‰“å¼€ QQ å°ç¨‹åºå¼€å‘è€…å·¥å…·è¿›è¡ŒæŸ¥çœ‹'
    },
    ["swan" /* SWAN */]: {
        OPEN: 'è¯·æ‰“å¼€ç™¾åº¦æ™ºèƒ½å°ç¨‹åºå¼€å‘è€…å·¥å…·è¿›è¡ŒæŸ¥çœ‹'
    },
    ["tt" /* TT */]: {
        OPEN: 'è¯·æ‰“å¼€å­—èŠ‚è·³åŠ¨å°ç¨‹åºå¼€å‘è€…å·¥å…·è¿›è¡ŒæŸ¥çœ‹'
    },
    ["jd" /* JD */]: {
        OPEN: 'è¯·æ‰“å¼€äº¬ä¸œå°ç¨‹åºå¼€å‘è€…å·¥å…·è¿›è¡ŒæŸ¥çœ‹'
    },
    ["quickapp" /* QUICKAPP */]: {
        OPEN: 'è¯·æŒ‰å¿«åº”ç”¨ç«¯å¼€å‘æµç¨‹ https://taro-docs.jd.com/taro/docs/quick-app.html è¿›è¡ŒæŸ¥çœ‹'
    }
};
const getServeSpinner = (() => {
    let spinner;
    return () => {
        if (!spinner) {
            spinner = ora(`å³å°†å¼€å§‹å¯åŠ¨ç¼–è¯‘ï¼Œè¯·ç¨ç­‰~`);
            spinner.start();
        }
        return spinner;
    };
})();
exports.getServeSpinner = getServeSpinner;
const printCompiling = () => {
    getServeSpinner().text = 'æ­£åœ¨ç¼–è¯‘...';
};
exports.printCompiling = printCompiling;
const printBuildError = (err) => {
    const message = err.message;
    const stack = err.stack;
    if (stack && message.indexOf('from UglifyJs') !== -1) {
        try {
            const matched = /(.+)\[(.+):(.+),(.+)\]\[.+\]/.exec(stack);
            if (!matched) {
                throw new Error('Using errors for control flow is bad.');
            }
            const problemPath = matched[2];
            const line = matched[3];
            const column = matched[4];
            console.log('Failed to minify the code from this file: \n\n', chalk_1.default.yellow(`\t${problemPath}:${line}${column !== '0' ? ':' + column : ''}`), '\n');
        }
        catch (ignored) {
            console.log('Failed to minify the bundle.', err);
        }
    }
    else {
        console.log((message || err) + '\n');
    }
    console.log();
};
exports.printBuildError = printBuildError;
const printSuccess = (buildAdapter) => {
    getServeSpinner().stopAndPersist({
        symbol: 'âœ… ',
        text: isFirst ? chalk_1.default.green(`ç¼–è¯‘æˆåŠŸï¼Œ${LOG_MAP[buildAdapter].OPEN}\n`) : chalk_1.default.green(`ç¼–è¯‘æˆåŠŸ\n`)
    });
};
const printWarning = () => {
    getServeSpinner().stopAndPersist({
        symbol: 'âš ï¸ ',
        text: chalk_1.default.yellow('ç¼–è¯‘è­¦å‘Š.\n')
    });
};
const printFailed = () => {
    getServeSpinner().stopAndPersist({
        symbol: 'ðŸ™…  ',
        text: chalk_1.default.red('ç¼–è¯‘å¤±è´¥.\n')
    });
};
const printWhenBeforeCompile = compiler => {
    compiler.hooks.beforeCompile.tap('taroBeforeCompile', filepath => {
        printCompiling();
    });
    return compiler;
};
const printWhenInvalid = compiler => {
    compiler.hooks.invalid.tap('taroInvalid', filepath => {
        printCompiling();
    });
    return compiler;
};
const printWhenFailed = compiler => {
    compiler.hooks.failed.tap('taroFailed', error => {
        printBuildError(error);
    });
    return compiler;
};
let isFirst = true;
const printWhenFirstDone = (compiler) => {
    compiler.hooks.done.tap('taroDone', stats => {
        if (isFirst) {
            isFirst = false;
            getServeSpinner().clear();
            console.log(chalk_1.default.gray('\nç›‘å¬æ–‡ä»¶ä¿®æ”¹ä¸­...\n'));
        }
    });
    return compiler;
};
const _printWhenDone = ({ verbose = false }, buildAdapter, compiler) => {
    compiler.hooks.done.tap('taroDone', stats => {
        const { errors, warnings } = formatMessages(stats);
        if (!stats.hasErrors() && !stats.hasWarnings()) {
            printSuccess(buildAdapter);
        }
        if (stats.hasErrors()) {
            printFailed();
            errors.forEach(e => console.log(e + '\n'));
            verbose && process.exit(1);
            return;
        }
        if (stats.hasWarnings()) {
            printWarning();
            warnings.forEach(w => console.log(w + '\n'));
        }
        verbose && console.log(stats.toString({
            colors: true,
            modules: false,
            children: false,
            chunks: false,
            chunkModules: false,
            warnings: verbose
        }) + '\n');
    });
    return compiler;
};
const printWhenDone = fp_1.partial(_printWhenDone, [{ verbose: false }]);
const printWhenDoneVerbosely = fp_1.partial(_printWhenDone, [{ verbose: true }]);
const bindDevLogger = (compiler, buildAdapter) => {
    console.log();
    fp_1.pipe(printWhenBeforeCompile, fp_1.partial(printWhenDone, [buildAdapter]), printWhenFailed, printWhenInvalid, printWhenFirstDone)(compiler);
    return compiler;
};
exports.bindDevLogger = bindDevLogger;
const bindProdLogger = (compiler, buildAdapter) => {
    console.log();
    fp_1.pipe(printWhenBeforeCompile, fp_1.partial(printWhenDoneVerbosely, [buildAdapter]), printWhenFailed)(compiler);
    return compiler;
};
exports.bindProdLogger = bindProdLogger;
